✨ Add support for deploying Cluster templates and ClusterClasses on Tilt (#6254)
kubernetes-sigs/cluster-api: docs/book/src/developer/tilt.md

**enable_providers** (Array[]String, default=['docker']): A list of the providers to enable. See [available providers](#available-providers)
for more details.

**flavor_dirs** (Map{String: Array[]String}, default={"docker": [

> "./test/infrastructure/docker/templates"]}): A map of providers to directories containing flavor templates. An example of the field is given below. See [Deploying a workload cluster](#deploying-a-workload-cluster) for how this is used.

Reviewer: ```suggestion
"./test/infrastructure/docker/templates"]}): A map of provider to directories containing flavor templates. An example of the field is given below. See [Deploying a workload cluster](#deploying-a-workload-cluster) for how this is used.
```

```json

> "flavor_dirs": {
>     "docker": [
>         "./test/infrastructure/docker/templates",
>         "<other-template-dir>"
>     ],
>     "azure": [
>         "<azure-template-dir>"
>     ],
>     "aws": [
>         "<aws-template-dir>"
>     ],
>     "gcp": [
>         "<gcp-template-dir>"
>     ]
> }

Reviewer 2: ```suggestion
flavor_dirs:
  docker:
  - ./test/infrastructure/docker/templates
  - <other-template-dir
  azure:
  - <azure-template-dir>"
  aws:
  - <aws-template-dir>"
  gcp:
  - <gcp-template-dir>
```
nit: to be consistent with the rest of the doc

Reviewer 2: Should we call it template_dirs? Usually we call this template relatively consistently (I think the only place where we call it flavor is when you select a cluster template on the clusterctl cli)

configs](https://github.com/kubernetes-sigs/cluster-api/tree/main/test/infrastructure/docker/examples) available.
These can be customized for your specific needs.

### Deploying a workload cluster

After your kind management cluster is up and running with Tilt, you can deploy a workload clusters in the Tilt web UI based off of YAML templates from specified directories. By default, templates are read from `./test/infrastructure/docker/templates`.

These deployment resources are found in the Tilt web UI under the label grouping `<provider>-cluster-templates` and `<provider>-clusterclasses` for each specified provider, i.e. `docker-cluster-templates` and `docker-clusterclasses`.

> The `<provider>-cluster-templates` category contains cluster templates, you can create a cluster by clicking "Create cluster" or the clockwise arrow icon ⟳. Note that each time a cluster template is deployed, it deploys a new workload cluster in addition to the existing ones. To delete clusters based off of a template, click on "Delete \<cluster-template\> cluster," and click on "Delete all workload clusters" to delete all workload clusters.

Reviewer 3: ```suggestion
The `<provider>-cluster-templates` category contains cluster templates, you can create a cluster by clicking "Create cluster" or the clockwise arrow icon ⟳. Note that each time a cluster template is deployed, it deploys a new workload cluster in addition to the existing ones. To delete all clusters based off of a template, click on "Delete \<cluster-template\> cluster," and click on "Delete all workload clusters" to delete all workload clusters.
```
nit

> The `<provider>-cluster-templates` category contains cluster templates, you can create a cluster by clicking "Create cluster" or the clockwise arrow icon ⟳. Note that each time a cluster template is deployed, it deploys a new workload cluster in addition to the existing ones. To delete clusteres based off of a template, click on "Delete \<cluster-template\> cluster," and click on "Delete all workload clusters" to delete all workload clusters.

Reviewer 2: ```suggestion
The `<provider>-cluster-templates` category contains cluster templates, you can create a cluster by clicking "Create cluster" or the clockwise arrow icon ⟳. Note that each time a cluster template is deployed, it deploys a new workload cluster in addition to the existing ones. To delete clusters based off of a template, click on "Delete \<cluster-template\> cluster," and click on "Delete all workload clusters" to delete all workload clusters.
```
nit

> The `<provider>-clusterclasses` category contains clusterclasses and you can create them by clicking on the "Create clusterclass" or the clockwise arrow icon ⟳ and delete them by clicking on "Delete clusterclass".

Reviewer 3: ```suggestion
The `<provider>-clusterclasses` category contains ClusterClass definitions and you can create them by clicking on the "Create clusterclass" or the clockwise arrow icon ⟳ and delete them by clicking on "Delete clusterclass".
```
nit

Variables in a cluster template are substituted with values from `kustomize_substitutions` in `tilt-settings.yaml`. The default substitutions are:

```yaml
kustomize_substitutions:
  NAMESPACE: default
  KUBERNETES_VERSION: v1.24.0
  CONTROL_PLANE_MACHINE_COUNT: 1
  WORKER_MACHINE_COUNT: 3
```

> Lastly, cluster template directories can be specified from the `template_directories` field in `tilt-settings.yaml`. See [tilt-settings fields](#tilt-settings-fields) for an example.

Reviewer 2: ```suggestion
Lastly, cluster template directories can be specified from the `template_dirs` field in `tilt-settings.yaml`. See [tilt-settings fields](#tilt-settings-fields) for an example.
```

configs](https://github.com/kubernetes-sigs/cluster-api/tree/main/test/infrastructure/docker/examples) available.
These can be customized for your specific needs.

### Deploying a workload cluster

After your kind management cluster is up and running with Tilt, you can deploy a workload clusters in the Tilt web UI based off of YAML templates from specified directories. By default, templates are read from `./test/infrastructure/docker/templates`.

> These deployment resources are found in the Tilt web UI under the label grouping `<provider>-flavors` and `<provider>-clusterclasses` for each specified provider, i.e. `docker-flavors` and `docker-clusterclasses`. Note that deploying a workload cluster from Tilt UI is also termed as flavor cluster deployment.

Reviewer 2: ```suggestion
These deployment resources are found in the Tilt web UI under the label grouping `<provider>-flavors` and `<provider>-clusterclasses` for each specified provider, i.e. `docker-flavors` and `docker-clusterclasses`. Note that deploying a workload cluster from Tilt UI is also termed as cluster flavor deployment.
```
? not sure

To be honest I'm hearing the term "flavor cluster" for the first time. Maybe just drop the last sentence, for me it sounds like we are establishing a new term which isn't really necessary here - and at least for me it is confusing

> The `<provider>-flavors` category contains flavor templates, you can create a cluster by clicking "Create flavor cluster" or the clockwise arrow icon ⟳. Note that each time a flavor is deployed, it deploys a new workload cluster in addition to the existing ones. To delete clusteres based off of a flavor, click on "Delete \<flavor\> cluster," and click on "Delete all flavor clusters" to delete all workload clusters.

Reviewer 2: ```suggestion
The `<provider>-flavors` category contains cluster templates, you can create a cluster by clicking "Create cluster" or the clockwise arrow icon ⟳. Note that each time a cluster is deployed, it deploys a new workload cluster in addition to the existing ones. To delete clusters based off of a flavor, click on "Delete \<flavor\> cluster," and click on "Delete all flavor clusters" to delete all workload clusters.
```
I think we usually call those cluster templates.

Reviewer 2: I would also suggest to call the tab in the Tilt UI `<provider>-clustertemplates`. I think this way we can avoid an awkward mix between flavor and cluster template, or call cluster templates flavors.

The `<provider>-clusterclasses` category contains clusterclasses and you can create them by clicking on the "Create clusterclass" or the clockwise arrow icon ⟳ and delete them by clicking on "Delete clusterclass".

> Variables in a flavor template are substituted with values from `kustomize_substitutions` in `tilt-settings.json`. The default substitutions are:

Reviewer 2: ```suggestion
Variables in a cluster template are substituted with values from `kustomize_substitutions` in `tilt-settings.yaml`. The default substitutions are:
```
